Refactor ImageCDN parsing to rely on HTML API instead of RegExps

dmsnell opened the pull request:
## Status

~This is a *work in progress* and *isn't tested* or verified.~ This has been reviewed, but the filters aren't tested because it's unclear what code might rely on them.

Due to the change in indentation the diff view is more associated with the actual changes if **ignoring whitespace**.

## Proposed changes:

The introduction of the HTML API into WordPress 6.2 offers a new method of matching and modifying HTML. In this patch we're replacing code that attempts to parse the input HTML and extract images that are direct children of an anchor ("A" tag), then read and modify them based on the values of their attributes and computed Photon properties.

In the previous code the `Image_CDN` class scanned the entire HTML document to generate a list of PREG image match objects, then iterated over those matches and performed string-replace operations on them.

Now the class does a pass from start to finish, visting each image tag along the way, and making the appropriate modifications. Extra care is taken to ensure that only images that are the single child of a link are matched.

In this change the values of the `tag` key in some of the filters has changed from the initial matched HTML snippet to the name of the image tag, which could be `IMG` or `AMP-IMG` or `AMP-ANIM`. An update to the Tag Processor or a custom sub-class thereof could provide the original HTML snippet and match the existing behavior, but that hasn't been done in this patch yet given the author's uncertainty about the use and value of those snippets.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

This mandates running on WordPress 6.2. If we prefer being able to run on older versions, we could also include a copy of the Tag Processor in the plugin.

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

Test suite should pass. Manual review is necessary though.
I don't even know what all the function I modified is supposed to do fully, so code auditing and review of the modifications is critical.

jeherve commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -279,9 +279,40 @@ public static function filter_photon_norezise_maybe_inject_sizes_api( $sizes, $a
+	 * Given a `class` attribute value, return the first class name with
+	 * the given prefix, excluding that prefix in the return value.
+	 *
+	 * @since {VERSION NUMBER}
```
We use `$$next-version$$` as the placeholder:

https://github.com/Automattic/jetpack/blob/3a3c4adc1c0913bac224ddb832713bd6101fe1ab/projects/packages/README.md#L107-L116

```suggestion
	 * @since $$next-version$$
```

dmsnell replied:
nice, thank you @jeherve!

jeherve commented on projects/packages/image-cdn/CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [0.2.4] - 2023-08-28
 ### Changed
 - Updated package dependencies. [#32605]
+- Rely on WordPress HTML API to parse HTML instead of Regex Patterns. [#43743]
```
We typically do not edit the changelog directly, since when you work on your PR you do not yet know when the change will be released / what version of the package will include your changes. 

Instead, you can run `jetpack changelog add packages/image-cdn` and follow the prompts to add your changelog entry. It will generate a changelog entry file that you can commit and that will be merged alongside your PR. When we next release an update to the Image CDN package, we'll use the different changelog entry files from the different PRs that were merged since the last release to:

1. Decide what version should be released, based on the severity declared in the different changelog entry files.
1. Generate the changelog for that version.

dmsnell replied:
is there a way to do this without running `jetpack` on my machine? I don't have `wp-cli` or Docker or know what the `jetpack` command is.

oskosk replied:
There is an introduction to the Jetpack development environment here https://github.com/Automattic/jetpack/blob/trunk/docs/development-environment.md#clone-the-repository and also a [Quick Start Guide here](https://github.com/Automattic/jetpack/blob/trunk/docs/quick-start.md)

jeherve replied:
> is there a way to do this without running jetpack on my machine?

If you'd rather not install [the Jetpack CLI tool](https://github.com/Automattic/jetpack/blob/trunk/tools/cli/README.md) (although it would be useful for other things like running the tests with `jetpack test packages/image-cdn php`), and have `composer` installed on your machine, you should be able to run the changelogger commands directly. The Changelog tool is [a composer package](https://packagist.org/packages/automattic/jetpack-changelogger) that can be installed in any project, and it is a dependency in the image-cdn package where you're working:
https://github.com/Automattic/jetpack/blob/51bb232b314a6cee550d2072d5bfe0b341d97c27/projects/packages/image-cdn/composer.json#L13

In practice, you'd run `composer install` in `projects/packages/image-cdn`, and then run `vendor/bin/changelogger add` to create a new changelog entry.

And of course you still have the option of doing it all manually by adding a file in the `projects/packages/image-cdn/changelog/` directory, using the expected format:
https://github.com/Automattic/jetpack/blob/750b20235e84757aeb6980708b62b88b480218ad/projects/packages/changelogger/README.md#adding-a-change-file

dmsnell replied:
thank you @jeherve !!!

jeherve replied:
Happy to help! And now you see the consequences of adding a changelog entry specifying a `patch` level change. GitHub checks expect you to bump the package's version to match that patch version increase, and give you another script you can run to do that automatically: `tools/fixup-project-versions.sh`

https://github.com/Automattic/jetpack/pull/32700/files

jeherve reviewed (commented):
Thank you for proposing this! This is a nice way to use `WP_HTML_Tag_Processor`, and something we can do now that we'll be requiring WordPress 6.2 (#31638).

I have not tested your PR, but left a couple of comments below about how things work in the monorepo, if that can help you discover how we usually do things :)

oskosk commented:
@haqadn  @Automattic/heart-of-gold can you make sure you keep an eye on this one?

dmsnell commented:
@oskosk @jeherve should I try and find someone to review this? are either of you happy with the proposed change?

dmsnell commented:
One note I should leave here is to ask about pulling in Core's `assertEqualMarkup()` function. I wanted to add it to the `BaseTestCase` but then I saw that's an external class and wasn't sure where to add it so that it would be available all around Jetpack.

The issue here is that the test output changes but in a semantically neutral way. I could rewrite the test so that it hardcodes the newer ordering of output, but that would leave this fragility in place for the future.

jeherve commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -336,297 +367,342 @@ public static function parse_dimensions_from_filename( $src ) {
+			/**
+			 * Allow specific images to be skipped by Photon.
+			 *
+			 * @TODO: Does this need to pass the full HTML of the image tag?
```
I looked at usage of this filter in other plugins in the WordPress.org plugin directory. While most plugins do not use this third parameter, there are a few that do, sometimes to look for a specific class name in the HTML. So I think we should keep this available if possible.

jeherve commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -336,297 +367,342 @@ public static function parse_dimensions_from_filename( $src ) {
+					 * Filter whether an image using an attachment ID in its class has to be uploaded to the local site to go through Photon.
 					 *
-					 * @see https://developer.wordpress.com/docs/photon/api/
+					 * @TODO: What is the point of passing $images and $index. Are they required?
```
I looked around but couldn't figure out where that filter is being used today. It was used in the past, but may not be used anymore today.

jeherve reviewed (commented):
> should I try and find someone to review this? are either of you happy with the proposed change?

Maybe @Automattic/heart-of-gold could take a look?

haqadn reviewed (dismissed):
The code changes look good to me and no problems found after testing. 👍🏼 

P.S.: I only worked on migrating from a Jetpack module to image-cdn package. I don't have prior knowledge of the implementation details on the core functionality.

dmsnell commented:
I will reconstruct the existing filter value before merge.
Would still love to have some thoughts on the `assertEquivalentMarkup()` method from Core, if there's a better place to put that.

kraftbj commented on projects/packages/image-cdn/tests/php/test_class.image_cdn.php:
```diff
@@ -178,6 +178,51 @@ protected function helper_remove_image_sizes() {
 		remove_image_size( 'jetpack_soft_oversized_after_upload' );
 	}
 
+	/**
```
I think these functions are okay here.

Personally, I would put them in the parent class at `class-image-cdn-attachment-test-case.php` but, it looks like this is the only (current) test that extends this particular class.

kraftbj commented:
Since we're late in the beta testing week, I'd say let's merge this next week after the branch cut so we get the longest amount of time to test it in the wild before the next dotorg release

dmsnell commented:
whoops, was I supposed to merge this? I'll try and rebase the branch and get it ready again without changing the code, then maybe we can merge next week or so.

thingalon commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -281,9 +281,40 @@ public static function filter_photon_norezise_maybe_inject_sizes_api( $sizes, $a
+			return null;
+		}
+
+		$search_pattern = sprintf( '#(?:^|[ \t\f\r\n])%s(?P<value>[^ \t\f\r\n]+)#', preg_quote( $prefix, null ) );
```
Style question: Is there a difference between `[\s]` and `[ \t\f\r\n]`? If not, would it be more readable using `\s`?

dmsnell replied:
good question! I'm so used to the HTML API where we've been very explicit about these that I didn't think about it, but [the PHP docs](https://www.php.net/manual/en/regexp.reference.escape.php) suggest that `\s` may be affected by locale options. I think it would be just as good to leave this as-is, though `\s` is unlikely to be a big problem in practice so if you prefer it I'm happy to make that change.

apparently it may include additional unicode characters representing whitespace if those locale rules are active, which would diverge from how the CSS class names are parsed in the browser.

thingalon replied:
Unicode can't let anything be simple, not even whitespace.
What you've written would be closer to the HTML spec, so let's stick with how you've done it.

thingalon commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -338,297 +369,342 @@ public static function parse_dimensions_from_filename( $src ) {
-				 * @param string $tag Image Tag (Image HTML output).
-				 */
-				if ( apply_filters( 'jetpack_photon_skip_image', false, $src, $tag ) ) {
+		// @TODO: Do "PICTURE" and "SOURCE" belong here as well?
```
Yes, I believe we should target these. However, I would do it in a separate PR to keep the reviewing/auditing easier, and less likely to produce surprises.

thingalon commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -338,297 +369,342 @@ public static function parse_dimensions_from_filename( $src ) {
+
+				// Detect WP registered image size from HTML class.
+				$class = $p->get_attribute( 'class' );
+				$size  = self::parse_class_name_prefixed_by( $class, 'size-' );
+				if ( null !== $size && false === $width && false === $height && 'full' !== $size && array_key_exists( $size, $image_sizes ) ) {
+					$width     = (int) $image_sizes[ $size ]['width'];
+					$height    = (int) $image_sizes[ $size ]['height'];
+					$transform = $image_sizes[ $size ]['crop'] ? 'resize' : 'fit';
+				} else {
+					unset( $size );
+				}
```
Very minor point, but can we skip calling `parse_class_name_prefixed_by` entirely if `width` or `height` are not `false`, skipping some regexes per-image where sizes are explicitly set?

dmsnell replied:
good call, why do more work than we need to?

thingalon reviewed (dismissed):
I just went through `filter_the_content` line-by-line, comparing its functionality to the original. Well done, I think you have nailed it!

I do have a couple of minor questions and suggestions - but absolutely none of them are blockers in my opinion so I'm giving this a ✅ on behalf of Heart of Gold.

Should we also remove `parse_images_from_html` which now only appears in its test suite?

dmsnell commented:
All, thanks for your feedback; I've _finally_ updated this PR and I think I have addressed everything. It now respects the old value for the `$tag` parameter in the `jetpack_photon_skip_image` filter.

haqadn reviewed (dismissed):
I fixed call to `WP_HTML_Tag_Processor` by using the correct case. Looks like the PR is ready to get merged.

dmsnell commented:
thanks @haqadn. sorry all I've been holding out on this so I can try and test it before merging. too much to do!

dmsnell committed:
Refactor ImageCDN parsing to rely on HTML API instead of RegExps (#32700)

The introduction of the HTML API into WordPress 6.2 offers a new method of
matching and modifying HTML. In this patch we're replacing code that attempts
to parse the input HTML and visit all images, then read and modify them based
on the values of their attributes and computed Photon properties.

In the previous code the `Image_CDN` class scanned the entire HTML document to
generate a list of PREG image match objects, then iterated over those matches
and performed string-replace operations on them.

Now the class does a pass from start to finish, visting each image tag along
the way, and making the appropriate modifications.

Co-authored-by: Adnan Haque <3737780+haqadn@users.noreply.github.com>
Co-authored-by: Brandon Kraft <public@brandonkraft.com>
Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>
Co-authored-by: Mark George <thingalon@gmail.com>
Co-authored-by: Osk <oskosk@users.noreply.github.com>

dmsnell committed:
Rearrange semantically equivalent test output to avoid false negatives.

No `assertEquivalentMarkup` exists yet, so this gets around that without
creating one.

dilirity commented:
I started looking into the failed tests, but I'm not sure what's going on. `str_starts_with` is returning incorrectly for some reason.

Any idea where that's loaded from?

dilirity committed:
Fix broken test

dilirity committed:
Remove unnecessary comment

dilirity committed:
Fix static analysis issues

dilirity committed:
Fix static analysis issue

dmsnell commented:
@dilirity - thanks for fixing those tests - not sure how I copy/pasted the original typo on `WP_HTML_Tag_Processor`, but I appreciate your help fixing it.

as far as I know it's ready to go, and it _was_ ready to go, but I'm not really sure how to test and vet it to the level of quality at which I would have preferred (since I was jumping in to unfamiliar code and proposing a change in a system I'm not working in).

if you think this is sound (and according to the tests and a code audit it seems ready) then it should be fine to merge. had I been in a place where I could thoroughly test or watch and react to failures post-merge I would have.

dilirity committed:
Merge branch 'trunk' into image-cdn/rely-on-html-api

dilirity committed:
Bump project version to 0.4.7-alpha

dilirity reviewed (commented):
<details><summary>My bad. I was using two different images. Ignore this. Though I'm not sure why only one of these images' width/height is 150 when both are set to use it.</summary>

So I did some testing and found something odd. Markup for resulted images below.

If you add an image via the Image block and use these settings:

![CleanShot 2024-08-27 at 17 50 36](https://github.com/user-attachments/assets/f9f5884e-e6ca-4792-b6ce-eddd5caa6257)

On the production version of Boost, it looks like this:

![CleanShot 2024-08-27 at 17 51 08](https://github.com/user-attachments/assets/98bf773f-8bf1-4d5e-a10e-156653475d89)

```html
<img decoding="async" width="150" height="150" src="https://i0.wp.com/honest-shark.jurassic.ninja/wp-content/uploads/2024/08/donnie-rosie-O7L3MrlSAHA-unsplash.jpg?resize=150%2C150&amp;ssl=1" alt="" class="wp-image-12" style="aspect-ratio:4/3;object-fit:cover" srcset="https://i0.wp.com/honest-shark.jurassic.ninja/wp-content/uploads/2024/08/donnie-rosie-O7L3MrlSAHA-unsplash-scaled.jpg?resize=150%2C150&amp;ssl=1 150w, https://i0.wp.com/honest-shark.jurassic.ninja/wp-content/uploads/2024/08/donnie-rosie-O7L3MrlSAHA-unsplash-scaled.jpg?zoom=2&amp;resize=150%2C150&amp;ssl=1 300w, https://i0.wp.com/honest-shark.jurassic.ninja/wp-content/uploads/2024/08/donnie-rosie-O7L3MrlSAHA-unsplash-scaled.jpg?zoom=3&amp;resize=150%2C150&amp;ssl=1 450w" sizes="(max-width: 150px) 100vw, 150px" data-recalc-dims="1">
```

However, this branch running Boost seems to ignore the size:

![CleanShot 2024-08-27 at 17 51 39](https://github.com/user-attachments/assets/6525e75b-3765-4eb1-835b-b0791c012808)

```html
<img fetchpriority="high" decoding="async" width="1707" height="2560" src="https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=150%2C150&amp;ssl=1" alt="" class="wp-image-14" style="aspect-ratio:4/3;object-fit:cover" srcset="https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?w=1707&amp;ssl=1 1707w, https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=200%2C300&amp;ssl=1 200w, https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=683%2C1024&amp;ssl=1 683w, https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=768%2C1152&amp;ssl=1 768w, https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=1024%2C1536&amp;ssl=1 1024w, https://i0.wp.com/extraordinary-gazelle.jurassic.ninja/wp-content/uploads/2024/08/filip-zrnzevic-QsWG0kjPQRY-unsplash-scaled.jpg?resize=1365%2C2048&amp;ssl=1 1365w" sizes="(max-width: 1000px) 100vw, 1000px">
```

I'm not sure if this is Boost related, but I don't think we've made any changes to Image CDN. Boost is running on the Free version with only `Critical CSS` and `Image CDN` enabled.

</details>

dilirity reviewed (dismissed):
Well, I couldn't find anything broken. Code-wise it is good and tests seem to be happy.

I think we can :shipit:

dilirity committed:
Merge branch 'trunk' into image-cdn/rely-on-html-api

dilirity committed:
Fix project version

dilirity reviewed (approved):
Approving after merge with trunk.
